File: challenger-backend/src/main/java/com/kameo/challenger/odb/UserODB.java
package com.kameo.challenger.odb;

import com.kameo.challenger.odb.api.IIdentity;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;
import lombok.ToString;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;
import java.util.List;

@Entity
@Data
@ToString(of=IIdentity.id_column)
@NoArgsConstructor
public class UserODB implements IIdentity {

    @Id
    @GeneratedValue(strategy= GenerationType.AUTO)
    long id;
    String login;
    String passwordHash;
    String salt;

    @NotNull
    String email;

    @NotNull
    @Enumerated
    UserStatus userStatus=UserStatus.ACTIVE;

    @Temporal(TemporalType.TIMESTAMP)
    Date suspendedDueDate;
    int failedLoginsNumber;

    @ManyToMany
    List<ChallengeODB> contracts;

    Long defaultChallengeContract;

    public UserODB(long id) {
        this.id=id;
    }

    // email is never null
    @Transient
    public String getLoginOrEmail() {
        return getLogin()!=null? getLogin(): getEmail();
    }

    public static UserODB ofEmail(String email) {
        UserODB u=new UserODB();
        u.setId(-1);
        u.setEmail(email);
        return u;
    }
}

File: challenger-backend/src/main/java/com/kameo/challenger/odb/UserODB.kt
package com.kameo.challenger.odb

import com.kameo.challenger.odb.api.IIdentity
import java.util.*
import javax.persistence.*


@Entity
data class UserODB( @Id
                    @GeneratedValue(strategy= javax.persistence.GenerationType.AUTO)
                    override val id:Long=0) : IIdentity {


    var login:String?=null;
    lateinit var email:String;
    lateinit var salt:String;
    lateinit var passwordHash:String;
    lateinit var userStatus:UserStatus;
    lateinit var suspendedDueDate:Date;
    var failedLoginsNumber:Int=0;
    var defaultChallengeContract:Long?=null;
    @ManyToMany(targetEntity = ChallengeODB::class)
    lateinit var contracts:MutableList<ChallengeODB>;

    fun getLoginOrEmail(): String {
        return login  ?:  email;

    }

     companion object {
         @JvmStatic fun ofEmail(email:String):UserODB
        {
            val u:UserODB = UserODB(-1);
            u.email=email;
            return u;
        }
    }

}


